Change higher levels to use trait instead of concrete NavigationController

Kandarp05 opened the pull request:
#602

Kandarp05 commented:
@ianthetechie  @Archdoog  Am i going in the right direction?

ianthetechie commented:
Yup, this is going the right direction! Thanks for the draft + ping!

From here, we'll need to make a way to use the logging version (maybe a boolean parameter to the function that returns these?) and, one of the classic unsolvable problems of computer science, better names ;)

Kandarp05 commented:
Added a boolean parameter should_record, also renamed `NavigationControllerTrait` to be `Navigator`, short and simple, let me know about your thoughts on this

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,6 +37,39 @@ pub struct NavigationController {
+/// Creates a new navigation controller for the given route and configuration.
+///
+/// It returns an Arc-wrapped trait object implementing `Navigator`.
+/// If `should_record` is true, it creates a controller that can record navigation events.
+/// Else it creates a regular controller without recording capabilities.
```
Minor edits:

```suggestion
/// If `should_record` is true, it creates a controller with event recording enabled.
```

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,6 +37,39 @@ pub struct NavigationController {
+        // For now, it just returns the regular controller
+        Arc::new(NavigationController { route, config })
+    } else {
+        // Creates a navigation controller without recording capabilities.
```
It's possible that other variations will be created later. The revised comment is a bit more future-proof.

```suggestion
        // Creates a normal navigation controller.
```

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,6 +37,39 @@ pub struct NavigationController {
+    should_record: bool,
+) -> Arc<dyn Navigator> {
+    if should_record {
+        // Creates a navigation controller with recording capabilities.
+        // For now, it just returns the regular controller
+        Arc::new(NavigationController { route, config })
```
```suggestion
        // Creates a navigation controller with a wrapper that records events.
        // TODO: Currently just returns the regular controller
        Arc::new(NavigationController { route, config })
```

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,6 +37,39 @@ pub struct NavigationController {
 
+/// Core interface for navigation functionalities.
+///
+/// This trait defines the essential operations for any navigation implementation
+/// allowing different navigation strategies (standard, recording)
+/// to be used interchangeably
```
Minor nits; the comment as written sounds very mechanical.

```suggestion
/// This trait defines the essential operations for a navigation state manager.
/// This lets us build additional layers (e.g. event logging)
/// around [`NavigationController`] a composable manner.
```

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -405,6 +438,20 @@ impl NavigationController {
     }
 }
 
+impl Navigator for NavigationController {
+    fn get_initial_state(&self, location: UserLocation) -> TripState {
+        self.get_initial_state(location)
+    }
+
+    fn advance_to_next_step(&self, state: &TripState) -> TripState {
+        self.advance_to_next_step(state)
+    }
+
+    fn update_user_location(&self, location: UserLocation, state: &TripState) -> TripState {
+        self.update_user_location(location, state)
+    }
+}
```
Rather than making methods that just call other methods by the same name, moving the implementation under the trait is more idiomatic.

(I'm actually surprised that this is legal; the signatures are the same, but yeah Rust traits can get a bit weird...)

ianthetechie reviewed (commented):
Looking good; mostly a bunch of minor nitpicks now.

* No update for Swift? I know you don't have a Mac, but this should be a trivial change along the lines of what you already did for Kotlin. We can clean up any errors if needed.
* The documentation (in the `guide` folder) needs to be updated, since the `FerrostarCore` constructors now require an extra argument.

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,16 +37,114 @@ pub struct NavigationController {
     }
+}
+
+/// Shared functionality for the navigation controller that is not exported by uniFFI.
```
Not your comment, but since it showed up in the diff... ;)

```suggestion
/// Shared functionality for the navigation controller that is not exported by UniFFI.
```

ianthetechie commented:
> Also the web is still using the old way, can you explain me how the JsNavigationController works and how I can update it to use the trait?

Yeah, we can walk through this on the call tomorrow as it's probably easier that way. (Probably not a bad side task for you to improve documentation around this too :D I don't think our GSoC student last year really documented this anywhere.)

ianthetechie reviewed (commented):
I think this is pretty good from my side. We can discuss JS tomorrow on the call. Until then, I'll leave this open so @Archdoog can have a quick look over too.

Archdoog commented:
I'd agree. Looking fine to me. Some considerations, but none are super important at this point 👍

ianthetechie commented on common/ferrostar/src/navigation_controller/mod.rs:
```diff
@@ -37,16 +37,114 @@ pub struct NavigationController {
+/// It returns an Arc-wrapped trait object implementing `Navigator`.
+/// If `should_record` is true, it creates a controller with event recording enabled.
+#[cfg_attr(feature = "uniffi", uniffi::export)]
+pub fn create_navigator(
```
Ooh, that's a fantastic idea! Either I'm forgetting some drawback of UniFFI that makes this impossible, or we should apply this everywhere ;) Let's give this a shot and even if it doesn't work out to expose in the UniFFI layer, we can probably make some more ergonomic APIs at the Swift and Kotlin level for better discoverability.
